### PR TITLE
tests: refactor expect in cidrToIP fcn in vmi_multus_test

### PR DESCRIPTION
**Pull request by jmguzik**
Signed-off-by: Jakub Guzik <jakubmguzik@gmail.com>

**What this PR does / why we need it**:
PR refactors one of the helper functions in tests/ to return an error to test case. I fixed only one expect right now just to get feedback if everything looks good, or it should be done another way

**Which issue(s) this PR fixes**
Related to #4101

**Release note**
```release-note
NONE
```

### CODE REVIEW

File: tests/vmi_multus_test.go [hunk 1]
```diff
@@ -1122,10 +1138,9 @@ var _ = Describe("[Serial]Macvtap", func() {
 	})
 })
 
-func cidrToIP(cidr string) string {
+func cidrToIP(cidr string) (string, error) {
 	ip, _, err := net.ParseCIDR(cidr)
-	Expect(err).ToNot(HaveOccurred(), "Should be able to parse IP and prefix length from CIDR")
-	return ip.String()
+	return ip.String(), err
```
Comment (phoracek) [inline tests/vmi_multus_test.go:1143]: Would you keep the error message here?
```go
fmt.Errorf("Failed to parse IP and prefix length from CIDR: %w", err)
```
Then we could drop the duplicated message from all the asserts and keep them as simple

```go
Expect(err).ToNot(HaveOccurred())
```

Comment (jmguzik) [reply tests/vmi_multus_test.go:1143]: I do not why, but I somehow understood error logging should be inside the test as well.
Now, this makes much more sense. Thanks, I will add a fix today.

Comment (phoracek) [reply tests/vmi_multus_test.go:1143]: There are no hard rules. To me it looks like a good compromise to keep it short while asserting outside the helper.

Oh and edit on the message - it should be "failed to parse IP from CIDR" - the function does not care about the prefix length

File: tests/vmi_multus_test.go [hunk 2]
```diff
@@ -1122,10 +1138,12 @@ var _ = Describe("[Serial]Macvtap", func() {
 	})
 })
 
-func cidrToIP(cidr string) string {
+func cidrToIP(cidr string) (string, error) {
 	ip, _, err := net.ParseCIDR(cidr)
-	Expect(err).ToNot(HaveOccurred(), "Should be able to parse IP and prefix length from CIDR")
-	return ip.String()
+	if err != nil {
+		fmt.Errorf("Failed to parse IP from CIDR: %w", err)
```
Comment (phoracek) [inline tests/vmi_multus_test.go:1144]: I believe this should be `return "", fmt.Errorf(...)`

Comment (jmguzik) [reply tests/vmi_multus_test.go]: Yes, I noticed that forgot to include one staged change, pushed now

**Comment by jmguzik** [comment]
@phoracek finally the whole CI went green, and the branch is rebased. If you find the time, let's put to end a miserable history of cidrToIP function :)
Side note: your tests are running for a several of hours, many interactions, and sometimes there are flaky (toggle) cases that are spoiling the whole build. While fixing them is necessary, maybe it would be worth to split test cases into smaller groups to contain them in a 1-2 hours run. On a project level, have you thought about it?

File: tests/vmi_multus_test.go [hunk 3]
```diff
@@ -1140,6 +1140,9 @@ var _ = Describe("[Serial]Macvtap", func() {
 
 func cidrToIP(cidr string) (string, error) {
 	ip, _, err := net.ParseCIDR(cidr)
+	if err != nil {
+		fmt.Errorf("Failed to parse IP from CIDR: %w", err)
```
Comment (EdDev) [inline tests/vmi_multus_test.go]: Could you please reason for this additional info? Why is the original error info not enough?

**Review by EdDev — CHANGES_REQUESTED** [review]
Thank you for the contribution.

It seems that this change includes multiple commits that just show the development process and not a logical split of changes.

Could you please squash them and re-push the change?

I think I have only one inline comment about the error text.

File: tests/vmi_multus_test.go [hunk 4]
```diff
@@ -847,31 +847,40 @@ var _ = Describe("[Serial]SRIOV", func() {
 		It("[test_id:3956]should connect to another machine with sriov interface over IPv4", func() {
 			cidrA := "192.168.1.1/24"
 			cidrB := "192.168.1.2/24"
+			ipA, errA := cidrToIP(cidrA)
+			Expect(errA).ToNot(HaveOccurred(), "should be able to parse IP and prefix length from CIDR A")
```
[the remark this answers is not in the document]
Comment (jmguzik) [reply tests/vmi_multus_test.go]: @EdDev it was removed after comments from @phoracek to include it in upper function

File: tests/vmi_multus_test.go (hunk 3 above)
Comment (jmguzik) [reply tests/vmi_multus_test.go]: @EdDev  I followed instructions from @phoracek to include this one. Please refer to resolved conversations.

File: tests/vmi_multus_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmguzik) [reply tests/vmi_multus_test.go]: @EdDev Regarding commits history, what you would like to do, squash them?

File: tests/vmi_multus_test.go (hunk 3 above)
Comment (jmguzik) [reply tests/vmi_multus_test.go]: @EdDev squashed everything and removed this message.

File: tests/vmi_multus_test.go [hunk 5]
```diff
@@ -881,31 +881,42 @@ var _ = Describe("[Serial]SRIOV", func() {
 		It("[test_id:3956]should connect to another machine with sriov interface over IPv4", func() {
 			cidrA := "192.168.1.1/24"
 			cidrB := "192.168.1.2/24"
+			ipA, errA := cidrToIP(cidrA)
```
Comment (EdDev) [inline tests/vmi_multus_test.go:884]: nit: There is no real need for a dedicated error variable, as simple `err` would work as well.
```
ipA, err := cidrToIP(cidrA)
...
ipB, err := cidrToIP(cidrB)
```

File: tests/vmi_multus_test.go [hunk 6]
```diff
@@ -881,31 +881,42 @@ var _ = Describe("[Serial]SRIOV", func() {
...
+					return libnet.PingFromVMConsole(vmi2, ipA)
 				}, 15*time.Second, time.Second).Should(Succeed())
 			})
 		})
 
 		It("[test_id:3957]should connect to another machine with sriov interface over IPv6", func() {
 			cidrA := "fc00::1/64"
 			cidrB := "fc00::2/64"
+			ipA, errA := cidrToIP(cidrA)
+			Expect(errA).ToNot(HaveOccurred())
```
Comment (EdDev) [inline tests/vmi_multus_test.go:907]: nit: Same, `err` is enough, no need for two distinct ones.

**Review by EdDev — COMMENTED** [review]
Thanks!
Some small nits inline.

Regarding the commit message, could you please use the text from the PR?
Or try something like this:
```
tests, multus: Change cidrToIP helper to return an error
```

**Comment by jmguzik** [comment]
@EdDev everything done

**Review by EdDev — APPROVED** [review]
Thank you!

**Comment by phoracek** [comment]
Thanks for the contribution @jmguzik !

**Pull request merged**